### DISCUSSION TITLE
Error due to parameter files : Mosaic LSM

### DISCUSSION THREAD

**Discussion by Gayathridevi123** [NASA-LIS/LISF#994] [Q&A] [answered]
Hi,

I'm trying with Mosaic LSM and when I set the
 Porosity data source: none

The model was giving error:-
setmosp -- reading soil files
Reading soil parameter table ./input/PARAMETERS/mos_parms/mosaic_soilparms_fao.txt
setmosp -- no porosity datasource selected
setmosp -- calling endrun
[ERR] endrun is being called
Then I have seen from https://github.com/NASA-LIS/LISF/blob/master/lis/surfacemodels/land/mosaic/setmosp.F90 line 263;
if porosity is set as "none" then end-run is called. I'm not able to find the porosity file for 1km,5km,10km, or 25km so I kept the Porosity data source: CONSTANT, and there was no error related to porosity.

But I'm getting error :-
**At line 129 of file ../surfacemodels/land/mosaic/mosdynp.F90 (unit = 16, file = './input/PARAMETERS/mos_parms/mosaic_monthlyvegparms_umd.txt')
Fortran runtime error: Bad real number in item 1 of list input
Error termination. Backtrace:
#0  0x2b6e28a2aac8 in read_real
at ../../../libgfortran/io/list_read.c:2026
https://github.com/NASA-LIS/LISF/pull/1  0x2b6e28a2cead in list_formatted_read_scalar
at ../../../libgfortran/io/list_read.c:2180
https://github.com/NASA-LIS/LISF/issues/2  0xaeb587 in ???
https://github.com/NASA-LIS/LISF/pull/3  0x59cdeb in ???
https://github.com/NASA-LIS/LISF/issues/4  0x5972e0 in ???
https://github.com/NASA-LIS/LISF/pull/5  0x5b87d1 in ???
https://github.com/NASA-LIS/LISF/pull/6  0xd3cbe2 in ???
https://github.com/NASA-LIS/LISF/issues/7  0x5cbc84 in ???
https://github.com/NASA-LIS/LISF/issues/8  0x4084b3 in ???
https://github.com/NASA-LIS/LISF/issues/9  0x2b6e296213d4 in ???
https://github.com/NASA-LIS/LISF/pull/10  0x408568 in ???
https://github.com/NASA-LIS/LISF/issues/11  0xffffffffffffffff in ???**

Given the Mosaic LSM parameters:- (from https://portal.nccs.nasa.gov/lisdata_pub/data/PARAMETERS/mos_parms/)
In the LIS user guide, these parameter files are specified and I tried with the same:-

**Mosaic vegetation parameter table:       ./input/PARAMETERS/mos_parms/mosaic_vegparms_umd.txt
Mosaic monthly vegetation parameter table: ./input/PARAMETERS/mos_parms/mosaic_monthlyvegparms_umd.txt
Mosaic soil parameter table:             ./input/PARAMETERS/mos_parms/mosaic_soilparms_fao.txt**

 mosaic monthly vegetation parameter table has values for 13 UMD vegtypes.  

When I changed the monthly vegetation parameter table as **"monvegpar_igbp.txt"** that error is gone but getting another error:-

**At line 111 of file ../surfacemodels/land/mosaic/setmosp.F90 (unit = 10, file = './input/PARAMETERS/mos_parms/mosaic_soilparms_fao.txt')
Fortran runtime error: Bad real number in item 1 of list input**
Error termination. Backtrace:
#0  0x2ab47023fac8 in read_real
	at ../../../libgfortran/io/list_read.c:2026
#1  0x2ab470241ead in list_formatted_read_scalar
	at ../../../libgfortran/io/list_read.c:2180

Can anyone tell which file combination works for Mosaic?

Looking forward to a solution.
****
Thank you and regards,
Gayathri

**Comment by dmocko** [comment, marked as answer]
Hi,

Can you please attach the ldt.config and lis.config files for Mosaic?

I'm confused about your settings, although I do have an idea what might be occurring.

Thanks,

David

**Reply by Gayathridevi123** [reply]
Hi David,

Thank you for your reply. ldt.config and lis.config files for Mosaic are attached.

[ldt.mosaic.txt](https://github.com/NASA-LIS/LISF/files/8129568/ldt.mosaic.txt)
[lis.mosaic.txt](https://github.com/NASA-LIS/LISF/files/8129570/lis.mosaic.txt)

Eagerly waiting for a solution.

Thank you for your help!
Regards,
Gayathri

**Reply by dmocko** [reply]
Hi Gayathri,

The porosity "none" error is a flag to remind the user that porosity data is needed for Mosaic.  The porosity (SMCMAX) is a look-up table in the mosaic_soilparms_fao.txt file, by soil texture type.  There are three soil layers in Mosaic, and there are separate porosity (SMCMAX) values for each layer.  It looks like your final error is related to you using "STATSGOFAO_Native" for the soil texture data source, but you are providing an FAO (only) soil parameter file.  You will need to change this config entry in your ldt.config file to "none", and then Mosaic will calculate the FAO texture based on the sand/clay/silt FAO files that you are already supplying.  After Mosaic has this texture calculated for each point, it will use the mosaic_soilparms_fao.txt look-up table.

You are getting the vegetation error on Line 129 because you are trying to use the "MODIS_Native" landcover data, but supplying the UMD look-up table.  MODIS has 20 vegetation types, while UMD has 13, so it's a read error when LIS is trying to read 20 values on each line, but the UMD file only has 13.  You will need to use the "UMD" landcover data.  "UMD" landcover datasets at various grid spacings are available here: https://portal.nccs.nasa.gov/lisdata_pub/data/PARAMETERS/UMD/

The Mosaic LSM does not support the MODIS vegetation cover.  There's actually a flag about this in the Mosaic code that will stop the run if "MODIS" is chosen (but we need to add a check for "MODIS_Native" as well).  See: https://github.com/NASA-LIS/LISF/blob/9df627a3445c89f748ff89b91f970e2ba556cc34/lis/surfacemodels/land/mosaic/setmosp.F90#L335  You need to run Mosaic with the UMD landcover, which will map the UMD types onto the 8 Mosaic vegetation types.  You can see more about the 8 Mosaic vegetation types in this document on pages 34 and 35: https://gmao.gsfc.nasa.gov/pubs/docs/Koster130.pdf

David

**Reply by Gayathridevi123** [reply]
Thank you David for your detailed reply. It helps me to get a better understanding of what's going on while running Mosaic LSM.

I changed my ldt.config as per your suggestions. Used the UMD landcover and landmask data and set the soil texture data source as "none". Also in the ldt and lis config files,  I mentioned **soil fraction data source: FAO**, after giving all these entries now there is no error due to the parameter files. But it is giving segmentation fault error.

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x2b632d5405cf in ???
#1  0x2b632d8b7256 in ???
#2  0x5ddc01 in ???
#3  0xae0680 in ???
#4  0x59c732 in ???
#5  0x598e20 in ???
#6  0x5d7ba0 in ???
#7  0xd3cc14 in ???
#8  0x5cbc84 in ???
#9  0x4084b3 in ???
#10  0x2b632d76f3d4 in ???
#11  0x408568 in ???
#12  0xffffffffffffffff in ???

I am wondering why this fault is happening... ldt.config, ldt log, lis config and lis.log files are attached. 

[ldt.config.txt](https://github.com/NASA-LIS/LISF/files/8139180/ldt.config.txt)
[ldtlog.txt](https://github.com/NASA-LIS/LISF/files/8139181/ldtlog.txt)
[lis.config.txt](https://github.com/NASA-LIS/LISF/files/8139182/lis.config.txt)
[lis_input_MOSAIC.d01.nc.txt](https://github.com/NASA-LIS/LISF/files/8139184/lis_input_MOSAIC.d01.nc.txt)
[lislog.txt](https://github.com/NASA-LIS/LISF/files/8139187/lislog.txt)

I request you to kindly check if anything is wrong with my config file settings.

Regards,
Gayathri

**Reply by dmocko** [reply]
Hi Gayathri,

I didn't see anything wrong with your config settings.

Can you re-configure/compile with debug mode, and see if there are more detailed error messages?

cd LISF/lis/make
gmake realclean
cd ../
./configure

When you see this prompt, enter "-1":

Optimization level (-3=strict checks with warnings, -2=strict checks, -1=debug, 0,1,2,3, default=2):  -1

Then please post the error that you see when you run.

David

**Reply by Gayathridevi123** [reply]
Hi David,

I recompiled with the debug mode. Now seeing errors :-

**Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
==== backtrace ====
 0 0x000000327b73fa26 __memcmp_sse4_1()  interp.c:0
 1 0x00000000001c6a7b _gfortran_compare_string()  ???:0
 2 0x000000000079888c __lis_timemgrmod_MOD_isalarmringing()  ???:0
 3 0x00000000012b0c76 mos_main_()  ???:0
 4 0x0000000000735119 lsmrun_()  ???:0
 5 0x000000000072f343 __lis_lsmmod_MOD_lis_lsm_run()  ???:0
 6 0x000000000079082e __lis_surfacemodelmod_MOD_lis_surfacemodel_run()  ???:0
 7 0x0000000001851e49 __retrospective_runmod_MOD_lis_run_retrospective()  ???:0
 8 0x0000000000778ab3 lisrun_()  ???:0
 9 0x00000000010bf49e MAIN__()  lisdrv.F90:0
10 0x00000000010bf584 main()  ???:0
11 0x000000327b61ed1d __libc_start_main()  ???:0
12 0x0000000000408619 _start()  ???:0
===================
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 0 on node nesac-hpc exited on signal 11 (Segmentation fault).**

Thank you for your time and consideration.

Kind regards,
Gayathri

**Reply by dmocko** [reply]
Hi,

I don't know what that error means.

Can you please copy your configure.lis file, which is in LISF/lis/make?

Are you able to run other LSMs successfully (such as our end-to-end testcase)?

David

**Reply by Gayathridevi123** [reply]
Hi,

Yes, I'm able to run test cases and simulations using Noah, Noah-MP, and RUC LSMs successfully. 
Here attaching the configure.lis file.
[configure.txt](https://github.com/NASA-LIS/LISF/files/8166509/configure.txt)

Thank you and regards,
Gayathri

**Reply by dmocko** [reply]
Hi,

Thanks for the configure.lis file.

I am taking a look into this, and will follow up.

David

**Reply by Gayathridevi123** [reply]
Thank you David for your support...

**Reply by dmocko** [reply]
Hi @Gayathridevi123 

I was able to reproduce your error.  My full error message is below:

```
[discover33:15458:0:15458] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
==== backtrace (tid:  15458) ====
 0  /usr/lib64/libucs.so.0(ucs_handle_error+0xe4) [0x2aaabaa7eda4]
 1  /usr/lib64/libucs.so.0(+0x2210c) [0x2aaabaa7f10c]
 2  /usr/lib64/libucs.so.0(+0x222c2) [0x2aaabaa7f2c2]
 3  ./LIS(_intel_fast_memcmp+0x1c) [0x251334c]
 4  ./LIS(for_cpstr+0x41) [0x24ab091]
 5  ./LIS() [0x9eaf26]
 6  ./LIS() [0x18c0953]
 7  ./LIS() [0x8c4e64]
 8  ./LIS() [0x8b8af6]
 9  ./LIS() [0x9d65fd]
10  ./LIS() [0x20e1de6]
11  ./LIS() [0x9b386e]
12  ./LIS() [0x1629557]
13  ./LIS() [0x414712]
14  /lib64/libc.so.6(__libc_start_main+0xf5) [0x2aaab253b725]
15  ./LIS() [0x414629]
=================================
forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source             
LIS                0000000002440D9A  Unknown               Unknown  Unknown
libpthread-2.22.s  00002AAAB2011C10  Unknown               Unknown  Unknown
LIS                000000000251334C  Unknown               Unknown  Unknown
LIS                00000000024AB091  Unknown               Unknown  Unknown
LIS                00000000009EAF26  lis_timemgrmod_mp        3256  LIS_timeMgrMod.F90
LIS                00000000018C0953  mos_main_                 382  mos_main.F90
LIS                00000000008C4E64  lsmrun_                   476  LIS_lsm_FTable.c
LIS                00000000008B8AF6  lis_lsmmod_mp_lis         554  LIS_lsmMod.F90
LIS                00000000009D65FD  lis_surfacemodelm         248  LIS_surfaceModelMod.F90
LIS                00000000020E1DE6  retrospective_run         198  retrospective_runMod.F90
LIS                00000000009B386E  lisrun_                   202  LIS_runmode_FTable.c
LIS                0000000001629557  MAIN__                     70  lisdrv.F90
LIS                0000000000414712  Unknown               Unknown  Unknown
libc-2.22.so       00002AAAB253B725  __libc_start_main     Unknown  Unknown
LIS                0000000000414629  Unknown               Unknown  Unknown
```

I was looking for the latter half of the error messages.  Did you not see them when you run?

Particularly, LIS crashed on line 3256 of the LIS_timeMgrMod routine.  But this is called from line 382 of mos_main.F90:

https://github.com/NASA-LIS/LISF/blob/af94ed88100f392c1b066d2aa743e0e64d27af44/lis/surfacemodels/land/mosaic/mos_main.F90#L382

Tracing this back further, I was able to determine that the issue is in mos_lsmMod.F90, where the alarms are set, but do not also set the alarms for each nest.  In mos_main.F90, the alarm check is also looking for the nest number, but it's not there, causing the crash.

Please change the below lines in mos_lsmMod.F90, and then re-compile:

```
--- a/lis/surfacemodels/land/mosaic/mos_lsmMod.F90
+++ b/lis/surfacemodels/land/mosaic/mos_lsmMod.F90
@@ -136,6 +136,7 @@ contains
 !EOP  
     integer :: n,i
     integer                 :: status
+    character*3         :: fnest
 
     allocate(mos_struc(LIS_rc%nnest))
     call readmoscrd()
@@ -165,11 +166,12 @@ contains
 !------------------------------------------------------------------------
        call LIS_update_timestep(LIS_rc, n, mos_struc(n)%ts)
 
-       call LIS_registerAlarm("Mosaic model alarm",&
+       write(fnest,'(i3.3)') n    
+       call LIS_registerAlarm("Mosaic model alarm "//trim(fnest),&
             mos_struc(n)%ts,&
             mos_struc(n)%ts)
 
-       call LIS_registerAlarm("Mosaic restart alarm",&
+       call LIS_registerAlarm("Mosaic restart alarm "//trim(fnest),&
             mos_struc(n)%ts,&
             mos_struc(n)%rstInterval)
```

The "a" code with the minus signs are the original code, and the "b" code with the plus signs are what you need to change.  Note that exact line numbers (e.g., Lines 136 and 165) are approximate, and may vary slightly in your code.

There was one other change that I had to make to get Mosaic to run successfully.  I was incorrect before about the porosity data.  Mosaic does need distributed porosity maps, instead of only setting the porosity in the look-up table by soil texture type.  I just copied the data that you need to the LIS data portal.  You need the 1KM data because the other soils data you are using is also 1KM:
    https://portal.nccs.nasa.gov/lisdata_pub/data/PARAMETERS/UMD/1KM/

porosity_FAO.L1.1gd4r.gz
porosity_FAO.L2.1gd4r.gz
porosity_FAO.L3.1gd4r.gz

Un-zip them after you download them, and also put in your "1KM" directory.

Next, you will need to edit your ldt.config file, and make sure they contain these config entries:

```
Porosity data source:            FAO
Porosity map:                    ./input/PARAMETERS/UMD/1KM/porosity_FAO
Porosity fill value:             0.30
```

LIS will auto-magically fill in the rest of the filenames for each of the Mosaic soil layers.

After you do both of these, you should be able to run.  Note that when you run, you may encounter errors in your lislog.0000 file about some fields not being defined for diagnostic output (see below, for example):

```
 [ERR] ECanop                                             field is not defined
 [ERR] for diagnostic output...
 [ERR] Please exclude it from the model output attributes table
 [ERR] Program stopping ..
 [ERR] endrun is being called
```

Please compare the mos_main.F90 routine against your MODEL_OUTPUT_TBL file so that:
1. the variable is defined for output by Mosaic in mos_main.F90 (if it is not, exclude this value from output by changing the first column after the variable name in the MODEL_OUTPUT_TBL file from "1" to "0"); and
2. that you have the same units in your MODEL_OUTPUT_TBL file as are available for output in mos_main.F90.  In my example above, my MODEL_OUTPUT_TBL file tries to output "ECanop" in units of "kg/m2".  However, the code in mos_main.F90 only has units of "kg/m2s" as an available units from this variable.  So you need to change the units column in MODEL_OUTPUT_TBL to get the output in units of "kg/m2s" instead.

Good luck,

David

**Reply by Gayathridevi123** [reply]
Hi David, 

**Thank you for all your hard work on this...** 

I didn't see other error messages when I run the model.  Anyway I will definitely try it and update soon. 

Many thanks,
Gayathri

**Reply by Gayathridevi123** [reply]
Hi David,

As you told I changed those lines in the mos_lsmMod.F90 file, recompiled LIS again and added the porosity file in ldt.config. Now I'm happy to say that Mosaic is running. 

Thank you very much for your help and support...

Kind regards,
Gayathri

**Comment by dmocko** [comment]
This bug fix was merged into the public-7.3 branch in #1007.
